File: windows/src/PalmierPro.Core/Analysis/WordCutPlanner.cs
using PalmierPro.Core.Editing;

namespace PalmierPro.Core.Analysis;

public enum CutAggressiveness
{
    Tight,
    Balanced,
    Loose,
}

public static class CutAggressivenessExtensions
{
    public static double KeptGapMs(this CutAggressiveness a) => a switch
    {
        CutAggressiveness.Tight => 60,
        CutAggressiveness.Loose => 320,
        _ => 150,
    };
}

public static class WordCutPlanner
{
    public readonly record struct Word(int StartFrame, int EndFrame, bool Selected);

    public static List<FrameRange> CutRanges(
        IReadOnlyList<Word> words, int clipStart, int clipEnd, int keepGapFrames)
    {
        var filtered = words.Where(w => w.EndFrame > w.StartFrame).ToList();
        if (clipEnd <= clipStart || filtered.Count == 0) return [];

        var half = Math.Max(0, keepGapFrames / 2);
        var ranges = new List<FrameRange>();
        var k = 0;
        while (k < filtered.Count)
        {
            if (!filtered[k].Selected) { k++; continue; }
            var l = k;
            while (l + 1 < filtered.Count && filtered[l + 1].Selected) l++;
            var left = k > 0 ? filtered[k - 1].EndFrame : clipStart;
            var right = l + 1 < filtered.Count ? filtered[l + 1].StartFrame : clipEnd;
            var runStart = filtered[k].StartFrame;
            var runEnd = filtered[l].EndFrame;
            var keepBefore = Math.Min(Math.Max(0, runStart - left), half);
            var keepAfter = Math.Min(Math.Max(0, right - runEnd), half);
            var start = Math.Max(clipStart, Math.Min(left + keepBefore, runStart));
            var end = Math.Min(clipEnd, Math.Max(runEnd, right - keepAfter));
            if (end > start) ranges.Add(new FrameRange(start, end));
            k = l + 1;
        }
        return RippleEngine.MergeRanges(ranges);
    }
}

File: Sources/PalmierPro/Transcription/WordCutPlanner.swift
import Foundation

enum CutAggressiveness: String, Sendable, CaseIterable {
    case tight, balanced, loose
    var keptGapMs: Double { switch self { case .tight: 60; case .balanced: 150; case .loose: 320 } }
}

enum WordCutPlanner {
    struct Word: Equatable {
        let startFrame: Int, endFrame: Int, selected: Bool
    }

    static func cutRanges(words: [Word], clipStart: Int, clipEnd: Int, keepGapFrames: Int) -> [FrameRange] {
        let words = words.filter { $0.endFrame > $0.startFrame }
        guard clipEnd > clipStart, !words.isEmpty else { return [] }
        let half = max(0, keepGapFrames / 2)
        var ranges: [FrameRange] = []
        var k = 0
        while k < words.count {
            guard words[k].selected else { k += 1; continue }
            var l = k
            while l + 1 < words.count, words[l + 1].selected { l += 1 }
            let left = k > 0 ? words[k - 1].endFrame : clipStart
            let right = l + 1 < words.count ? words[l + 1].startFrame : clipEnd
            let runStart = words[k].startFrame, runEnd = words[l].endFrame
            let keepBefore = min(max(0, runStart - left), half)
            let keepAfter = min(max(0, right - runEnd), half)
            let start = max(clipStart, min(left + keepBefore, runStart))
            let end = min(clipEnd, max(runEnd, right - keepAfter))
            if end > start { ranges.append(FrameRange(start: start, end: end)) }
            k = l + 1
        }
        return RippleEngine.mergeRanges(ranges)
    }
}
